commons-mongo/src/main/scala/com/avsystem/commons/mongo/BsonReaderInput.scala

```diff
@@ -67,10 +70,60 @@ class BsonReaderObjectInput private(br: BsonReader) extends ObjectInput {
 
   override def hasNext: Boolean = it.hasNext
   override def nextField(): BsonReaderFieldInput = it.next()
+  private[mongo] def peek(): BsonReaderFieldInput = it.peek()
 }
+
 object BsonReaderObjectInput {
   def startReading(br: BsonReader): BsonReaderObjectInput = {
     br.readStartDocument()
     new BsonReaderObjectInput(br)
   }
 }
+
+/**
+  * Modifies iteration order - moves "_id" field to last position if it was first.
+  * This is a workaround for [[MongoWithIdPolymorphicGenCodec]] where objType needs to be read first.
+  */
+private[commons] class ObjectWithIdBsonInput[ID](objectInput: BsonReaderObjectInput)(implicit idCodec: GenCodec[ID]) extends ObjectInput {
+  private val id: Opt[ID] =
+    if (objectInput.hasNext && objectInput.peek().fieldName == "_id") Opt.some(idCodec.read(objectInput.nextField()))
+    else Opt.empty
+
+  private var idToRead = id.isDefined
+
+  override def hasNext: Boolean = objectInput.hasNext || idToRead
+  override def nextField(): FieldInput =
+    if (objectInput.hasNext) {
+      objectInput.nextField()
+    }
+    else if (idToRead) {
+      idToRead = false
+      new BsonValueFieldInput("_id", id.get)
+    } else {
+      throw new NoSuchElementException
+    }
+}
+
+private[commons] class BsonValueFieldInput[T](name: String, value: T) extends BsonInput with FieldInput {
+  override def fieldName: String = name
+  override def inputType: InputType = InputType.Simple
+
+  override def readNull(): Null = read[Null]
+  override def readString(): String = read[String]
+  override def readBoolean(): Boolean = read[Boolean]
+  override def readInt(): Int = read[Int]
+  override def readLong(): Long = read[Long]
+  override def readDouble(): Double = read[Double]
+  override def readBinary(): Array[Byte] = read[Array[Byte]]
+  override def readList(): ListInput = read[ListInput]
+  override def readObjectId(): ObjectId = read[ObjectId]
+
+  override def readObject(): ObjectInput = throw new UnsupportedOperationException
```

Comment:
wsparcie dla idkow ktore sa dokumentami bedzie ciezko napisac, dla mojego use casu jest to kompletnie niepotrzebne, zastanawiam sie, czy tego tak nie zostawic.